### PR TITLE
Support CrossRef in CSL styles

### LINKED ISSUE #7378: CSL styles do not take CrossRef fields into account
Bug report from the [Forum](https://discourse.jabref.org/t/cross-references-in-preview/2517)

>I have the following entries in my bibliography:

> 
 ```
@InCollection{smit2021,
 author = {Smith, Bob},
title = {An article},
 pages = {1-10},
 crossref = {jone2021},
 }
```
> 
```
 @Book{jone2021,
editor = {Jones, John},
 publisher = {Great Publisher},
 title = {A book},
 year = {2021},
address = {Somewhere},
 }
```
> 
> When using the default entry preview (“Customized preview style”), the preview includes information from the cross-referenced entry:
> 
> InCollection (smit2021)
> 
> Smith, Bob
> An article
> 2021
> 
> Jones, John (Ed.)
> Great Publisher: Somewhere
> p. 1-10
> 
> But if I try other preview styles, there is no information from the cross-referenced entry. For example, this is what I get from the APA 7th edition style (and I’ve seen the same thing in other styles):
> 
> Smith, B. An article (pp. 1–10).
> 
> Is it only possible to show the information from the cross-referenced entry in the customized preview style?
> 
> Thanks

FIX:
Should  be getResolvedFieldOrAlias
https://github.com/JabRef/jabref/blob/94f4c1ebba44e0bfceb6867938b116f0cadbbcce/src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java#L98-L99

### PR DESCRIPTION
**Pull request by Siedlerchr**
Fixes #7378 and also provides groundwork for any fix regarding #8372 

Removed copy of CSL to ASCIIDOC, RTF and XSL_FO as these are no longer supported

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by Siedlerchr** [comment]
Failing fetcher test are not relevant (RFC the url format changed)

File: src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java [hunk 1]
```diff
@@ -86,18 +88,26 @@ private void initialize(String newStyle, CitationStyleOutputFormat newFormat) th
...
+        private static CSLItemData bibEntryToCSLItemData(BibEntry bibEntry, BibDatabase bibDatabase) {
+            bibDatabase = bibDatabase != null ? bibDatabase : new BibDatabase();
+
             String citeKey = bibEntry.getCitationKey().orElse("");
             BibTeXEntry bibTeXEntry = new BibTeXEntry(new Key(bibEntry.getType().getName()), new Key(citeKey));
 
             // Not every field is already generated into latex free fields
             RemoveNewlinesFormatter removeNewlinesFormatter = new RemoveNewlinesFormatter();
+
+            // adapted from TexBibEntriesResolver
```
Comment (koppor) [inline src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java:105]: Please add another comment why the code is necessary. Maybe, it is `// Use information of referenced entry`.

Moreover, I think, the code is wrong for BibLaTeX. It should be externalized to a utility class - or to the BibEntry class? TypedBibEntry?

We should implement the content of Appendix B of the documentation: 

grafik

File: src/main/java/org/jabref/logic/citationstyle/CitationStyleGenerator.java [hunk 2]
```diff
@@ -24,40 +25,40 @@ private CitationStyleGenerator() {
...
     protected static String generateCitation(BibEntry entry, CitationStyle style) {
-        return generateCitation(entry, style.getSource(), CitationStyleOutputFormat.HTML);
+        return generateCitation(entry, style.getSource());
     }
 
     /**
-     * Generates a Citation based on the given entry and style
+     * Generates a Citation based on the given entry and style with a default bib database
      *
      * @implNote the citation is generated using JavaScript which may take some time, better call it from outside the main Thread
```
Comment (koppor) [inline src/main/java/org/jabref/logic/citationstyle/CitationStyleGenerator.java:39]: This line can be removed

File: src/main/java/org/jabref/logic/citationstyle/CitationStyleGenerator.java [hunk 3]
```diff
@@ -24,40 +25,40 @@ private CitationStyleGenerator() {
...
      */
     protected static String generateCitation(BibEntry entry, String style) {
-        return generateCitation(entry, style, CitationStyleOutputFormat.HTML);
+        return generateCitation(entry, style, CitationStyleOutputFormat.HTML, new BibDatabase());
```
Comment (koppor) [inline src/main/java/org/jabref/logic/citationstyle/CitationStyleGenerator.java:42]: HTML still seems to be supported, which is good.

File: src/test/java/org/jabref/logic/citationstyle/CitationStyleGeneratorTest.java [hunk 4]
```diff
@@ -66,80 +59,72 @@ void testHtmlFormat() {
...
-        String style = CitationStyle.getDefault().getSource();
-        CitationStyleOutputFormat format = CitationStyleOutputFormat.RTF;
-
-        String actualCitation = CitationStyleGenerator.generateCitation(entry, style, format);
+        String actualCitation = CitationStyleGenerator.generateCitation(entry, style, format, new BibDatabase());
         assertEquals(expectedCitation, actualCitation);
     }
 
     @Test
+    @Disabled("Currently citeproc does not handle number field correctly https://github.com/JabRef/jabref/issues/8372")
```
Comment (koppor) [inline src/test/java/org/jabref/logic/citationstyle/CitationStyleGeneratorTest.java:67]: This is strange, the test string reads correctly. Did your changes really make this test not passing?

**Review by koppor — CHANGES_REQUESTED** [review]

File: src/test/java/org/jabref/logic/citationstyle/CitationStyleGeneratorTest.java (hunk 4 above)
Comment (Siedlerchr) [reply src/test/java/org/jabref/logic/citationstyle/CitationStyleGeneratorTest.java:67]: First of all, the whole class was disabled and my changes did not touch this:
This is what it outputs for the test entry: 
`[1]B. Smith, B. Jones, and J. Williams, “Title of the test entry,” BibTeX Journal, vol. 34, no. 3, Art. no. 3, Jul. 2016, doi: 10.1001/bla.blubb.
`

File: src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java (hunk 1 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java:105]: Ah okay, I see we have already implemented this. It's hidden in the resolvedFieldOrAlias methods.
Problem is that I need to get all Fields (including the empty ones) for the type and therefore I need a bibdatabasecontext. Not sure if I am able to pass it down

File: src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java [hunk 5]
```diff
@@ -86,18 +88,26 @@ private void initialize(String newStyle, CitationStyleOutputFormat newFormat) th
...
 
         private final List<BibEntry> data = new ArrayList<>();
+        private BibDatabase bibDatabase;
 
         /**
          * Converts the {@link BibEntry} into {@link CSLItemData}.
          */
-        private static CSLItemData bibEntryToCSLItemData(BibEntry bibEntry) {
+        private static CSLItemData bibEntryToCSLItemData(BibEntry bibEntry, BibDatabase bibDatabase) {
+            bibDatabase = bibDatabase != null ? bibDatabase : new BibDatabase();
```
[the remark this answers is not in the document]
Comment (calixtus) [reply src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java:97]: We do, but only in some very special cases, where an if/than/else would produce way too much overhead.

[the remark this answers is not in the document]
Comment (calixtus) [reply src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java:97]: Btw, overwriting a method argument? Sure you wanna do that?

File: CHANGELOG.md [hunk 6]
```diff
@@ -13,14 +13,15 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
...
+
 ### Fixed
 
 - We fixed an issue where an exception could occur when saving the preferences [#7614](https://github.com/JabRef/jabref/issues/7614)
 
 ### Removed
 
-
-
+- We removed the option to copy CSL Citation styles data as XSL_FO, ASCIIDOC, and RTF as these are no longer supported in the external library used for processing the styles [#7378](https://github.com/JabRef/jabref/issues/7378)
```
[the remark this answers is not in the document]
Comment (Siedlerchr) [reply CHANGELOG.md]: Did not find any info on this.

File: src/main/java/org/jabref/gui/maintable/RightClickMenu.java [hunk 7]
```diff
@@ -106,11 +106,8 @@ private static Menu createCopySubMenu(ActionFactory factory,
...
             Menu copyCitationMenu = factory.createMenu(StandardActions.COPY_CITATION_MORE);
-            copyCitationMenu.getItems().addAll(
-                    factory.createMenuItem(StandardActions.COPY_CITATION_TEXT, new CopyCitationAction(CitationStyleOutputFormat.TEXT, dialogService, stateManager, clipBoardManager, taskExecutor, previewPreferences)),
-                    factory.createMenuItem(StandardActions.COPY_CITATION_RTF, new CopyCitationAction(CitationStyleOutputFormat.RTF, dialogService, stateManager, clipBoardManager, taskExecutor, previewPreferences)),
-                    factory.createMenuItem(StandardActions.COPY_CITATION_ASCII_DOC, new CopyCitationAction(CitationStyleOutputFormat.ASCII_DOC, dialogService, stateManager, clipBoardManager, taskExecutor, previewPreferences)),
-                    factory.createMenuItem(StandardActions.COPY_CITATION_XSLFO, new CopyCitationAction(CitationStyleOutputFormat.XSL_FO, dialogService, stateManager, clipBoardManager, taskExecutor, previewPreferences)));
+            copyCitationMenu.getItems().add(
```
Comment (koppor) [inline src/main/java/org/jabref/gui/maintable/RightClickMenu.java:109]: Can we add `COPY_CITATION_HTML`? - This is IMHO a proper replacement for RTF, ASCII_DOC, XSLFO

    factory.createMenuItem(StandardActions.COPY_CITATION_HTML, new CopyCitationAction(CitationStyleOutputFormat.HTML, dialogService, stateManager, clipBoardManager, taskExecutor, previewPreferences)),

**Review by koppor — CHANGES_REQUESTED** [review]
Small wish added - otherwise LGTM!

File: src/test/java/org/jabref/logic/citationstyle/CitationStyleGeneratorTest.java [hunk 8]
```diff
@@ -33,113 +41,176 @@ void testIgnoreCarriageReturnNewLine() {
...
+        String citation = CitationStyleGenerator.generateCitation(entry, CitationStyle.getDefault().getSource(), CitationStyleOutputFormat.TEXT, bibDatabaseContext, bibEntryTypesManager);
+        assertEquals(expected, citation);
+    }
+
+    @Test
+    void testBiblatexWithIssueAndNumber() {
+        BibEntry entry = new BibEntry(StandardEntryType.Article);
+        entry.setField(StandardField.AUTHOR, "Last, First and\nDoe, Jane");
+        entry.setField(StandardField.ISSUE, "7");
+        entry.setField(StandardField.ISSUE, "28");
```
Comment (ThiloteE) [inline src/test/java/org/jabref/logic/citationstyle/CitationStyleGeneratorTest.java:188]: Is this a typo? Issue x2. Above it says "withIssueAndNumber".

File: src/test/java/org/jabref/logic/citationstyle/CitationStyleGeneratorTest.java [hunk 9]
```diff
@@ -33,113 +41,176 @@ void testIgnoreCarriageReturnNewLine() {
...
+
+        String citation = CitationStyleGenerator.generateCitation(entry, CitationStyle.getDefault().getSource(), CitationStyleOutputFormat.TEXT, bibDatabaseContext, bibEntryTypesManager);
+        assertEquals(expected, citation);
+    }
+
+    @Test
+    void testBiblatexWitNumber() {
+        BibEntry entry = new BibEntry(StandardEntryType.Article);
+        entry.setField(StandardField.AUTHOR, "Last, First and Doe, Jane");
+        entry.setField(StandardField.NUMBER, "7");
```
Comment (ThiloteE) [inline src/test/java/org/jabref/logic/citationstyle/CitationStyleGeneratorTest.java:171]: At the top you used "3" for number. Maybe here too?

**Review by ThiloteE — COMMENTED** [review]
@Siedlerchr How can i help you? Sorry, at least ~90% of this pull request is something i would need to look up, though. It is above my current skill level.

File: src/main/java/org/jabref/gui/maintable/RightClickMenu.java (hunk 7 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/maintable/RightClickMenu.java:109]: See two lines above, HTML has been already available

Comment (koppor) [reply src/main/java/org/jabref/gui/maintable/RightClickMenu.java]: OK - I did not expect that being present in a different menu.

Can we merge it into one? The current sub menu has only one item!

grafik

New:

grafik

File: src/test/java/org/jabref/logic/citationstyle/CitationStyleGeneratorTest.java (hunk 8 above)
Comment (Siedlerchr) [reply src/test/java/org/jabref/logic/citationstyle/CitationStyleGeneratorTest.java:188]: Oh thanks for spotting! This was not intended.

Comment (Siedlerchr) [reply src/test/java/org/jabref/logic/citationstyle/CitationStyleGeneratorTest.java]: As you can now see at the tests, the bibtex number gets now rendered as Art no.

**Pull request merged**